pkg/descheduler/controllers/migration/arbitrator/sort.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2023 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package arbitrator
+
+import (
+	"context"
+	"math"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/descheduler/fieldindex"
+	utilclient "github.com/koordinator-sh/koordinator/pkg/util/client"
+)
+
+const (
+	JobKind = "Job"
+)
+
+// SortJobsByPod sort a SortFn that sorts PodMigrationJobs by their Pods, including priority, QoS.
+func SortJobsByPod(sorter func(pods []*corev1.Pod)) SortFn {
+	return func(jobs []*v1alpha1.PodMigrationJob, podOfJob map[*v1alpha1.PodMigrationJob]*corev1.Pod) []*v1alpha1.PodMigrationJob {
+		var pods []*corev1.Pod
+		jobOfPod := map[*corev1.Pod]*v1alpha1.PodMigrationJob{}
+		rankOfJob := map[*v1alpha1.PodMigrationJob]int{}
+		for _, job := range jobs {
+			pod := podOfJob[job]
+			if pod == nil {
+				rankOfJob[job] = math.MinInt
+				continue
+			}
+			pods = append(pods, pod)
+			jobOfPod[pod] = job
+		}
+
+		// using PodSorter to sort
+		sorter(pods)
+
+		for i, pod := range pods {
+			rankOfJob[jobOfPod[pod]] = i
+		}
+
+		sort.SliceStable(jobs, func(i, j int) bool {
+			return rankOfJob[jobs[i]] < rankOfJob[jobs[j]]
+		})
+		return jobs
+	}
+}
+
+// SortJobsByCreationTime returns a SortFn that stably sorts PodMigrationJobs by create time.
+func SortJobsByCreationTime() SortFn {
+	return func(jobs []*v1alpha1.PodMigrationJob, podOfJob map[*v1alpha1.PodMigrationJob]*corev1.Pod) []*v1alpha1.PodMigrationJob {
+		// calculate the time interval between the creation of migration job and the current for each job.
+		timeMap := map[*v1alpha1.PodMigrationJob]int64{}
+		for _, job := range jobs {
+			timeMap[job] = job.GetCreationTimestamp().Unix()
+		}
+		// perform stable sorting.
+		sort.SliceStable(jobs, func(i, j int) bool {
+			return timeMap[jobs[i]] > timeMap[jobs[j]]
+		})
+		return jobs
+	}
+}
+
+// SortJobsByMigrationStatus returns a SortFn that.
+func SortJobsByMigrationStatus(c client.Client) SortFn {
+	return func(podMigrationJobs []*v1alpha1.PodMigrationJob, podOfJob map[*v1alpha1.PodMigrationJob]*corev1.Pod) []*v1alpha1.PodMigrationJob {
+		// get owner of jobs
+		owners := map[types.UID]*metav1.OwnerReference{}
```

Comment:
It's unnecessary to define the variable `owners`.

Reply:
I want to get job owners of PMJs in waiting queue. Then check if there are some migrating PMJs under the owner.